### PR TITLE
ICU-22251 Move sprintf to snprintf.

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22251
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [ ] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Review by richgillam — APPROVED** [review]
This all looks good.  Looks mechanical, except for the calculation of the buffer size.  I didn't review all of those carefully, but the ones I did all looked right.

**Review by richgillam — APPROVED** [review]
New stuff looks good too.

**Review by roubert — DISMISSED** [review]
I really don't think that you should be hardcoding any of these numbers, that wouldn't get us any real improved memory safety, it would just move the places where things could go wrong.

Instead, existing code like this:

```
char tmp[16];
sprintf(tmp, "%d", value);
```

Should be replaced with new code like this:

```
char tmp[16];
snprintf(tmp, sizeof tmp, "%d", value);
```

Or, when that's not possible, some kind of logic / argument passing that ensures that the value used is ultimately calculated by the compiler from the actual buffer size.

**Comment by FrankYFTang** [comment]
> I really don't think that you should be hardcoding any of these numbers, that wouldn't get us any real improved memory safety, it would just move the places where things could go wrong.
> 
> Instead, existing code like this:
> 
> ```
> char tmp[16];
> sprintf(tmp, "%d", value);
> ```
> 
> Should be replaced with new code like this:
> 
> ```
> char tmp[16];
> snprintf(tmp, sizeof tmp, "%d", value);
> ```
> 
> Or, when that's not possible, some kind of logic / argument passing that ensures that the value used is ultimately calculated by the compiler from the actual buffer size.

**Comment by FrankYFTang** [comment]
> Should be replaced with new code like this:
> 
> ```
> char tmp[16];
> snprintf(tmp, sizeof tmp, "%d", value);
> ```



> ```
> 
> ```
> 
> 
>     
>   
> 
> Should be replaced with new code like this:
> 
> 
> ```
> char tmp[16];
> snprintf(tmp, sizeof tmp, "%d", value);
> ```

DONE

File: icu4c/source/test/cintltst/udatatst.c [hunk 1]
```diff
@@ -1157,7 +1157,7 @@ static void TestICUDataName()
     }
 
     /* Only major number is needed. */
-    sprintf(expectDataName, "%s%d%c",
+    snprintf(expectDataName, 20, "%s%d%c",
```
Comment (richgillam) [inline icu4c/source/test/cintltst/udatatst.c:1160]: Is there a reason why you didn't use `sizeof` here?

File: icu4c/source/test/cintltst/udatatst.c [hunk 2]
```diff
@@ -1626,7 +1626,7 @@ TestSwapCase(UDataMemory *pData, const char *name,
 
 static void U_CALLCONV
 printErrorToString(void *context, const char *fmt, va_list args) {
-    vsprintf((char *)context, fmt, args);
+    vsnprintf((char *)context, 100, fmt, args);
```
Comment (richgillam) [inline icu4c/source/test/cintltst/udatatst.c:1629]: Is there a reason you didn't use `sizeof` here?

File: icu4c/source/test/intltest/alphaindextst.cpp [hunk 3]
```diff
@@ -698,7 +698,7 @@ void AlphabeticIndexTest::TestJapaneseKanji() {
     int32_t overflowIndex = immIndex->getBucketCount() - 1;
     for(int32_t i = 0; i < UPRV_LENGTHOF(kanji); ++i) {
         char msg[40];
-        sprintf(msg, "kanji[%d]=U+%04lX in overflow bucket", (int)i, (long)kanji[i]);
+        snprintf(msg, 40, "kanji[%d]=U+%04lX in overflow bucket", (int)i, (long)kanji[i]);
```
Comment (richgillam) [inline icu4c/source/test/intltest/alphaindextst.cpp:701]: `sizeof(msg)`?

**Review by richgillam — DISMISSED** [review]
Looks good, and I think Fredrik's recommendation is a good one and makes the code more readable and bulletproof.  But I think you missed a couple spots...

File: icu4c/source/test/cintltst/udatatst.c (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/udatatst.c:1160]: somehow I missed this

File: icu4c/source/test/cintltst/udatatst.c (hunk 2 above)
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/udatatst.c:1629]: this is much hader to get the size and I cannot simply use sizeof here

**Comment by FrankYFTang** [comment]
PTAL

File: icu4c/source/test/cintltst/udatatst.c (hunk 2 above)
Comment (richgillam) [reply icu4c/source/test/cintltst/udatatst.c:1629]: Well, it's unit-test code, so I'm not terribly concerned.

**Review by richgillam — APPROVED** [review]
It all looks fine.  Cut and print.

**Comment by FrankYFTang** [comment]
> I really don't think that you should be hardcoding any of these numbers, that wouldn't get us any real improved memory safety, it would just move the places where things could go wrong.
> 
> Instead, existing code like this:
> 
> ```
> char tmp[16];
> sprintf(tmp, "%d", value);
> ```
> 
> Should be replaced with new code like this:
> 
> ```
> char tmp[16];
> snprintf(tmp, sizeof tmp, "%d", value);
> ```
> 
> Or, when that's not possible, some kind of logic / argument passing that ensures that the value used is ultimately calculated by the compiler from the actual buffer size.

done

**Pull request merged**